### PR TITLE
Limit serving of insecure metrics by allowing configurable IP

### PR DESCRIPTION
**Pull request by swatisehgal**
Currently we are serving insecure metrics on all IPv4 routable addresses on the local machine (0.0.0.0).

In this PR, we make the metric IP configurable so that in order to ensure that we listen for insecure metrics port only on one IP (localhost by default).

### CODE REVIEW

File: cmd/resource-topology-exporter/main.go [hunk 1]
```diff
@@ -87,7 +87,7 @@ func main() {
 	if err != nil {
 		klog.Fatalf("failed to setup metrics: %v", err)
 	}
-	err = metricssrv.Setup(parsedArgs.RTE.MetricsMode, metricssrv.NewConfig(parsedArgs.RTE.MetricsPort, parsedArgs.RTE.MetricsTLSCfg))
+	err = metricssrv.Setup(parsedArgs.RTE.MetricsMode, metricssrv.NewConfig(parsedArgs.RTE.MetricsIP, parsedArgs.RTE.MetricsPort, parsedArgs.RTE.MetricsTLSCfg))
```
Comment (ffromani) [inline cmd/resource-topology-exporter/main.go:90]: can we call the variable `MetricsAddress` ? (everywhere)

File: pkg/config/defaults.go [hunk 2]
```diff
@@ -39,6 +39,7 @@ func SetDefaults(pArgs *ProgArgs) {
 	pArgs.RTE.AddNRTOwnerEnable = true
 	pArgs.RTE.MetricsMode = metricssrv.ServingDefault
 	pArgs.RTE.MetricsPort = metricssrv.PortDefault
+	pArgs.RTE.MetricsIP = metricssrv.IPDefault
```
Comment (ffromani) [inline pkg/config/defaults.go:42]: Likewise `AddressDefault` perhaps

File: pkg/config/environ.go [hunk 3]
```diff
@@ -63,4 +63,7 @@ func FromEnv(pArgs *ProgArgs) {
 	if pArgs.RTE.MetricsPort == 0 {
 		pArgs.RTE.MetricsPort = metricssrv.PortFromEnv()
 	}
+	if pArgs.RTE.MetricsIP == "" {
+		pArgs.RTE.MetricsIP = metricssrv.IPFromEnv()
+	}
```
Comment (ffromani) [inline pkg/config/environ.go:68]: the fact we get the metric port from env is a historical accident, but let's be consistent

File: pkg/metrics/server/setup.go [hunk 4]
```diff
@@ -30,6 +30,7 @@ import (
 
 const (
 	PortDefault = 2112
+	IPDefault   = "127.0.0.1"
```
Comment (ffromani) [inline pkg/metrics/server/setup.go:33]: let's use "0.0.0.0" for backward compatibility

File: pkg/metrics/server/setup.go [hunk 5]
```diff
@@ -60,19 +61,21 @@ type TLSConfig struct {
 }
 
 type Config struct {
+	IP   string
```
Comment (ffromani) [inline pkg/metrics/server/setup.go:64]: Likewise, let's call this `Address`

**Review by ffromani — COMMENTED** [review]
LGTM

minor comments

File: pkg/config/environ.go (hunk 3 above)
Comment (swatisehgal) [reply pkg/config/environ.go:68]: Ack, Thanks!

File: pkg/metrics/server/setup.go (hunk 4 above)
Comment (swatisehgal) [reply pkg/metrics/server/setup.go:33]: Ack, will update.

**Comment by ffromani** [comment]
the e2e failure is curl failing with error 35 which is SSL connection error:
https://serverfault.com/questions/606135/curl-35-ssl-connect-error

**Comment by ffromani** [comment]
curl being obsolete is unlikely the culprit. Let's add `-v` to all the curl commandlines (`curl -v -L ...`) to have more debugging infos

**Comment by swatisehgal** [comment]
> curl being obsolete is unlikely the culprit. Let's add `-v` to all the curl commandlines (`curl -v -L ...`) to have more debugging infos

Ack, will do now.

**Comment by swatisehgal** [comment]
> the e2e failure is curl failing with error 35 which is SSL connection error: https://serverfault.com/questions/606135/curl-35-ssl-connect-error

```
  [FAILED] failed exec command on pod. pod="default/resource-topology-exporter-ds-cnkcw"; cmd=["curl" "-v" "-L" "https://127.0.0.1:2112/metrics"]; err=failed to run command [curl -v -L https://127.0.0.1:2112/metrics]: command terminated with exit code 35; stderr=""
  Unexpected error:
      <*fmt.wrapError | 0xc0002882c0>: 
      failed to run command [curl -v -L https://127.0.0.1:2112/metrics]: command terminated with exit code 35
      {
          msg: "failed to run command [curl -v -L https://127.0.0.1:2112/metrics]: command terminated with exit code 35",
          err: <exec.CodeExitError>{
              Err: <*errors.errorString | 0xc0000694c0>{
                  s: "command terminated with exit code 35",
              },
              Code: 35,
          },
      }
  occurred
```
I tried a few things suggested in the link you recommended as well as a few others but didn't have much luck. 
Will investigate this tomorrow with a fresh mind.

File: test/e2e/rte/metrics.go [hunk 6]
```diff
@@ -120,7 +123,7 @@ var _ = ginkgo.Describe("[RTE][Monitoring] metrics", func() {
 			}
 			rteContainerName, err := e2ertepod.FindRTEContainerName(rtePod)
 			gomega.Expect(err).ToNot(gomega.HaveOccurred())
-			cmd := []string{"curl", "-k", "-L", fmt.Sprintf("https://127.0.0.1:%d/metrics", metricsPort)}
+			cmd := []string{"curl", "-k", "-L", fmt.Sprintf("https://%s:%d/metrics", MetricsAddress, metricsPort)}
```
Comment (ffromani) [inline test/e2e/rte/metrics.go]: this seems to work fine

File: test/e2e/rte/metrics.go [hunk 7]
```diff
@@ -139,7 +142,7 @@ var _ = ginkgo.Describe("[RTE][Monitoring] metrics", func() {
 			}
 			rteContainerName, err := e2ertepod.FindRTEContainerName(rtePod)
 			gomega.Expect(err).ToNot(gomega.HaveOccurred())
-			cmd := []string{"curl", "-L", fmt.Sprintf("http://127.0.0.1:%d/metrics", metricsPort)}
+			cmd := []string{"curl", "-L", fmt.Sprintf("https://%s:%d/metrics", MetricsAddress, metricsPort)}
```
Comment (ffromani) [inline test/e2e/rte/metrics.go]: here we want to test over http (plain text) but you changed the protocol part to "httpS" (emphasis added), which is wrong in the context of this test

File: test/e2e/rte/metrics.go [hunk 8]
```diff
@@ -173,7 +176,7 @@ var _ = ginkgo.Describe("[RTE][Monitoring] metrics", func() {
 			rteContainerName, err := e2ertepod.FindRTEContainerName(rtePod)
 			gomega.Expect(err).ToNot(gomega.HaveOccurred())
 
-			cmd := []string{"curl", "-L", fmt.Sprintf("http://127.0.0.1:%d/metrics", metricsPort)}
+			cmd := []string{"curl", "-L", fmt.Sprintf("https://%s:%d/metrics", MetricsAddress, metricsPort)}
```
Comment (ffromani) [inline test/e2e/rte/metrics.go]: same here, this test wants to access over plain http so let's not change the URL part. Or, like above, let's change the test accordingly (but we will need to excercise the plaintext option somehow)

**Review by ffromani — COMMENTED** [review]
weird that `-v` doesn't give us more output, but I think I found the culprit anyway

File: test/e2e/rte/metrics.go (hunk 7 above)
Comment (swatisehgal) [reply test/e2e/rte/metrics.go]: Ah, good catch!!

File: test/e2e/rte/metrics.go (hunk 8 above)
Comment (swatisehgal) [reply test/e2e/rte/metrics.go]: Ack, Thanks!

File: pkg/metrics/server/setup.go [hunk 9]
```diff
@@ -130,6 +133,14 @@ func PortFromEnv() int {
 	return port
 }
 
+func IPFromEnv() string {
```
Comment (ffromani) [inline pkg/metrics/server/setup.go:136]: please rename `AddressFromEnv` for consistency

**Review by ffromani — COMMENTED** [review]
almost there, let's fix a minor error issue which survived the last round of updates and we can merge (and begin the backport dance)

File: pkg/metrics/server/setup.go (hunk 9 above)
Comment (swatisehgal) [reply pkg/metrics/server/setup.go:136]: On it.

**Review by ffromani — COMMENTED** [review]
thanks, LGTM!

**Pull request merged**